### PR TITLE
Add support for search attributes

### PR DESCRIPTION
**Pull request by MarcosCela**
**What changed?**

This adds support for search attributes:

Defines two new startup flags to pre-register search attributes:  `search-attributes-type` & `search-attributes-key`. If the user gives values for these flags (multiple values must be separated by a comma), then temporalite will initialize the custom search attributes as defined in the user input.
In addition to this, we also give the ability to pre-register search attributes on a temporalite `TestServer`.

**How did you test it?**

Tests are included.

### CODE REVIEW

**Comment by cretz** [comment]
While I see that this adds attributes to the cluster metadata, do search attributes actually work with this? I was under the impression elastic search was required. Can `workflow.UpsertSearchAttributes` be used here?

**Comment by MarcosCela** [comment]
> While I see that this adds attributes to the cluster metadata, do search attributes actually work with this? I was under the impression elastic search was required. Can `workflow.UpsertSearchAttributes` be used here?

Hi! Yes search attributes "work" without ElasticSearch. You can use:

`workflowUpsertSearchAttributes' or even:

```go
client.StartWorkflowOptions{
			TaskQueue:        "hello_world",
			SearchAttributes: map[string]interface{}{"test": "test-value"},
		},
```

And they will run just fine. I am far from an expert but my tests show that you only need to pre-register the search attribute name and type before working with them. In addition to that, searching by search attributes will fail.

File: cmd/temporalite/main.go [hunk 1]
```diff
@@ -64,6 +67,14 @@ func buildCLI() *cli.App {
...
 			ArgsUsage: " ",
 			Flags: []cli.Flag{
+				&cli.StringSliceFlag{
+					Name:  searchAttrKey,
+					Usage: "Optional search attributes keys that will be registered at startup. If there are multiple keys, concatenate them and separate by ,",
+				},
+				&cli.StringSliceFlag{
+					Name:  searchAttrType,
+					Usage: "Optional search attributes types that will be registered at startup. If there are multiple keys, concatenate them and separate by ,",
+				},
```
[the remark this answers is not in the document]
Comment (MarcosCela) [reply cmd/temporalite/main.go:77]: For sure! The first one looks _better_ to me!

File: temporaltest/server_test.go [hunk 2]
```diff
@@ -112,3 +113,56 @@ func BenchmarkRunWorkflow(b *testing.B) {
...
+	// Ensure that the search attributes returned by the server
+	// have the search attribute defined in the test
+	value, ok := searchAttrs.Keys["test"]
+	if ok {
+		if value != enums.INDEXED_VALUE_TYPE_TEXT {
+			t.Fatal(fmt.Sprintf("search attribute was defined and present, but the value did not match. Expected %s, but got %s", enums.INDEXED_VALUE_TYPE_TEXT, value.String()))
+		}
+	} else {
+		t.Fatal("search attribute was defined, but not returned by the server")
+	}
```
[the remark this answers is not in the document]
Comment (MarcosCela) [reply temporaltest/server_test.go:167]: moved it to a new test!

File: temporaltest/server_test.go [hunk 3]
```diff
@@ -112,3 +113,56 @@ func BenchmarkRunWorkflow(b *testing.B) {
...
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Get workflow result
+	var result string
+	if err := wfr.Get(context.Background(), &result); err != nil {
+		t.Fatal(err)
+	}
```
[the remark this answers is not in the document]
Comment (MarcosCela) [reply temporaltest/server_test.go:148]: Done, took me a while to figure out decoding the proto message!

File: dbutil.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	"go.temporal.io/server/common/resolver"
+)
+
+type searchAttributesHelper struct {
+	db sqlplugin.DB
+}
+
+func (s *searchAttributesHelper) Close() error {
+	return s.db.Close()
+}
```
[the remark this answers is not in the document]
Comment (MarcosCela) [reply dbutil.go]: fixed!

**Pull request closed without merging**